clients/drcachesim/scheduler/scheduler_dynamic.cpp

```diff
@@ -179,7 +186,11 @@ typename scheduler_tmpl_t<RecordType, ReaderType>::stream_status_t
 scheduler_dynamic_tmpl_t<RecordType, ReaderType>::swap_in_input(output_ordinal_t output,
```

Comment:
Possible concern: Maybe I missed this in the previous PR review: Do we guarantee swap_in_input and swap_output_input events to be properly ordered. If not, you can imagine the derived subclasses seeing swap_in for an input before it was swap_out (and races too).

Reply:
Yes they are ordered.  It would be strange if they weren't right?  Added a comment "    // This is properly ordered with respect to swap_out_input()."

Reply:
I'm guessing your reply was to note the proper ordering between swap_out_input and swap_in_input calls made by the same thread in the same invocation to set_cur_input.

But my concern involves multiple threads with an ordering like: T1 swaps out some input X, T2 swaps in X, T2 calls swap_in_input, T1 calls swap_out_input.

Reply:
By "thread" and "T" I assume you actually mean "outputs" or "cores".  If you look at the swap_out_input() code in _dynamic, it says:

    // Now that the caller has updated the outgoing input's fields (we assert that
    // cur_output was changed above), add it to the ready queue (once on the queue others
    // can see it and pop it off).

That is the key: no other output can steal an input if it's not in the queue, and the currently executing input is not in the queue.  So T2 cannot swap in X until _dynamic adds it back to the queue.  Thus swap_in cannot happen before swap_out for an input X.  But this is all in the subclass's control: replay has strict ordering for recreating; fixed has no migrations; and dynamic has the queue and its lock as just described.  The swap interface need make no guarantees: it is up to the subclasses.

Reply:
Ack, thanks.